frappe/public/js/frappe/form/multi_select_dialog.js

```diff
@@ -2,86 +2,205 @@ frappe.ui.form.MultiSelectDialog = class MultiSelectDialog {
 	constructor(opts) {
 		/* Options: doctype, target, setters, get_query, action, add_filters_group, data_fields, primary_action_label */
 		Object.assign(this, opts);
-		var me = this;
-		if (this.doctype != "[Select]") {
-			frappe.model.with_doctype(this.doctype, function () {
-				me.make();
-			});
+		this.for_select = this.doctype == "[Select]";
+		if (!this.for_select) {
+			frappe.model.with_doctype(this.doctype, () => this.init());
 		} else {
-			this.make();
+			this.init();
 		}
 	}
 
-	make() {
-		let me = this;
+	init() {
 		this.page_length = 20;
 		this.start = 0;
-		let fields = this.get_primary_filters();
+		this.fields = this.get_fields();
+
+		this.make();
+	}
 
-		// Make results area
-		fields = fields.concat([
-			{ fieldtype: "HTML", fieldname: "results_area" },
+	get_fields() {
+		const primary_fields = this.get_primary_filters();
+		const result_fields = this.get_result_fields();
+		const data_fields = this.get_data_fields();
+		const child_selection_fields = this.get_child_selection_fields();
+
+		return [...primary_fields, ...result_fields, ...data_fields, ...child_selection_fields];
+	}
+
+	get_result_fields() {
+		const show_next_page = () => {
+			this.start += 20;
+			this.get_results();
+		};
+		return [
 			{
-				fieldtype: "Button", fieldname: "more_btn", label: __("More"),
-				click: () => {
-					this.start += 20;
-					this.get_results();
-				}
+				fieldtype: "HTML", fieldname: "results_area"
+			},
+			{
+				fieldtype: "Button", fieldname: "more_btn",
+				label: __("More"), click: show_next_page.bind(this)
 			}
-		]);
+		];
+	}
+
+	get_data_fields() {
+		if (this.data_fields && this.data_fields.length) {
+			// Custom Data Fields
+			return [
+				{ fieldtype: "Section Break" },
+				...this.data_fields
+			];
+		} else {
+			return [];
+		}
+	}
 
-		// Custom Data Fields
-		if (this.data_fields) {
-			fields.push({ fieldtype: "Section Break" });
-			fields = fields.concat(this.data_fields);
+	get_child_selection_fields() {
+		const fields = [];
+		if (this.child_selection_mode && this.child_doctype) {
```

Comment:
```suggestion
		if (this.allow_child_item_selection && this.child_fieldname) {
```

- reading **child_selection_mode** it feels like only child items can be selected.
- `child_fieldname` instead of `child_doctype` because there can be more than one child table pointing to the same child doctype in a doctype.